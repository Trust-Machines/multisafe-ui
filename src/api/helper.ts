import {ClarityAbi} from '@stacks/transactions';
import {NETWORK} from '@trustmachines/multisafe-contracts';

export const validateSafeAbi = (abi: ClarityAbi) => {
    const publicFns = [
        'add-owner',
        'remove-owner',
        'set-threshold',
        'submit',
        'confirm',
        'revoke'
    ]

    for (let p of publicFns) {
        if (!abi.functions.find(x => x.name === p && x.access === 'public')) {
            return false;
        }
    }

    return true;
}

export const makeTxUrl = (txId: string, network: NETWORK) => {
    return `https://explorer.stacks.co/txid/${txId}?chain=${network}`;
}