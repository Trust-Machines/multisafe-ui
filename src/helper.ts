import {ClarityAbi} from '@stacks/transactions';
import {NETWORK} from '@trustmachines/multisafe-contracts';
import BigNumber from 'bignumber.js';

export const validateSafeAbi = (abi: ClarityAbi) => {
    const publicFns = [
        'add-owner',
        'remove-owner',
        'set-min-confirmation',
        'submit',
        'confirm',
        'revoke'
    ]

    for (let p of publicFns) {
        if (!abi.functions.find(x => x.name === p && x.access === 'public')) {
            return false;
        }
    }

    return true;
}

export const makeTxUrl = (txId: string, network: NETWORK) => {
    return `https://explorer.stacks.co/txid/${txId}?chain=${network}`;
}

export const formatUnits = (value: BigNumber | string, decimals: number): BigNumber => {
    const bn = typeof value === 'string' ? new BigNumber(value) : value;
    return bn.dividedBy(10 ** decimals);
}

export const parseUnits = (value: string, decimals: number): BigNumber => {
    return new BigNumber(value).multipliedBy(10 ** decimals);
}