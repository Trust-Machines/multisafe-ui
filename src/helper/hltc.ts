import { script, Network, networks, payments } from 'bitcoinjs-lib';
import {
    bytesToHex,
    IntegerType,
    intToHexString,
    hexToBytes,
    utf8ToBytes,
} from 'micro-stacks/common';
import { hashSha256 } from 'micro-stacks/crypto-sha';

export interface HTLC {
    hash: Uint8Array;
    senderPublicKey: Uint8Array;
    recipientPublicKey: Uint8Array;
    expiration?: number;
    swapper: IntegerType;
}

export type BufferType = Buffer | Uint8Array;

export function numberToLE(num: IntegerType, length = 4) {
    const hexBE = intToHexString(num, length);
    let le = '';
    // reverse the buffer
    for (let i = 0; i < length; i++) {
        le += hexBE.slice(-2 * (i + 1), -2 * i || length * 2);
    }
    return le;
}

export function numberToLEBytes(num: IntegerType, length = 4) {
    return hexToBytes(numberToLE(num, length));
}

export const CSV_DELAY = 500;
export const CSV_DELAY_BUFF = script.number.encode(CSV_DELAY);
export const CSV_DELAY_HEX = CSV_DELAY_BUFF.toString('hex');

export function encodeExpiration(expiration?: number): Buffer {
    return typeof expiration === 'undefined' ? CSV_DELAY_BUFF : script.number.encode(expiration);
}

export function htlcASM({ hash, senderPublicKey, recipientPublicKey, expiration, swapper }: HTLC) {
    const swapperHex = numberToLE(swapper);
    return `
  ${swapperHex} OP_DROP
	OP_IF
    OP_SHA256 ${bytesToHex(hash)}
    OP_EQUALVERIFY
		${bytesToHex(recipientPublicKey)}
	OP_ELSE
		${encodeExpiration(expiration).toString('hex')}
		OP_CHECKSEQUENCEVERIFY
		OP_DROP
		${bytesToHex(senderPublicKey)}
	OP_ENDIF
	OP_CHECKSIG`
        .replace(/\s+/g, ' ')
        .trim();
}

export function generateHTLCScript(htlc: HTLC) {
    const asm = htlcASM(htlc);
    const output = script.fromASM(asm);
    return output;
}

export function generateHTLCAddress(htlc: HTLC, _network?: Network) {
    const network = _network || networks.regtest;
    const script = generateHTLCScript(htlc);
    // const payment = payments.p2wsh({ redeem: { output: script, network }, network });
    // const payment = payments.p2sh({ , network });
    // const hash = ripemd160(Uint8Array.from(script));
    // const payment = payments.p2sh({ output: Buffer.from(hash), network });
    const payment = payments.p2sh({ redeem: { output: script, network }, network });
    return payment;
}

export function reverseBuffer(buffer: BufferType): Uint8Array {
    if (buffer.length < 1) return buffer;
    let j = buffer.length - 1;
    let tmp = 0;
    for (let i = 0; i < buffer.length / 2; i++) {
        tmp = buffer[i];
        buffer[i] = buffer[j];
        buffer[j] = tmp;
        j--;
    }
    return Uint8Array.from(buffer);
}

export function getScriptHash(output: BufferType): Uint8Array {
    const uintOutput = Uint8Array.from(output);
    const hash = hashSha256(uintOutput);
    const reversed = reverseBuffer(Buffer.from(hash));
    return reversed;
}

export function hexPadded(hex: string) {
    const bytes = hex.length % 2 ? `0${hex}` : hex;
    return bytes;
}

export function secretToHash(secret: string) {
    const bytes = utf8ToBytes(secret);
    return hashSha256(bytes);
}
