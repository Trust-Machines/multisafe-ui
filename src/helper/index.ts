import BigNumber from 'bignumber.js';
import {DEPLOYERS, NETWORK} from '@trustmachines/multisafe-contracts';
import {StacksNetwork} from '@stacks/network';
import {contractPrincipalCV, validateStacksAddress} from '@stacks/transactions';
import {NETWORKS} from '../constants';
import {escapeRegExp} from '../util';
import {TransactionType} from '../types';

export const formatUnits = (value: BigNumber | string, decimals: number): BigNumber => {
    const bn = typeof value === 'string' ? new BigNumber(value) : value;
    return bn.dividedBy(10 ** decimals);
}

export const parseUnits = (value: string, decimals: number): BigNumber => {
    return new BigNumber(value).multipliedBy(10 ** decimals);
}

export const checkDecimalAmount = (input: string) => {
    const re = RegExp(`^\\d*(?:\\\\[.])?\\d*$`);

    if (!re.test(escapeRegExp(input))) {
        return false;
    }

    if (input.indexOf('.') !== -1) {
        return input.split('.')[1].length <= 6;
    }

    return true;
}

export const checkAmount = (input: string) => {
    return RegExp(/^(\s*|\d+)$/).test(input)
}

export const getStacksNetwork = (n: NETWORK): StacksNetwork => {
    return NETWORKS[n];
}

export const detectTransactionType = (executor: string): TransactionType => {
    const exec = executor.replace("'", '');
    for (const d of DEPLOYERS) {
        if (`${d}.add-owner` === exec) {
            return 'add-owner';
        }

        if (`${d}.remove-owner` === exec) {
            return 'remove-owner';
        }

        if (`${d}.set-threshold` === exec) {
            return 'set-threshold';
        }

        if (`${d}.transfer-stx` === exec) {
            return 'transfer-stx';
        }

        if (`${d}.transfer-sip-009` === exec) {
            return 'transfer-sip-009';
        }

        if (`${d}.transfer-sip-010` === exec) {
            return 'transfer-sip-010';
        }
    }

    return null;
}

export const contractPrincipalCVFromString = (s: string) => {
    const [a, b] = s.split(".");
    return contractPrincipalCV(a, b);
}

export const transformNftUri = (uri: string, nftId: string) => {
    const u = uri.replace('{id}', nftId);

    if (u.startsWith('ipfs://')) {
        return u.replace(/ipfs:\/\/(ipfs\/)?/, 'https://gateway.ipfs.io/ipfs/');
    }

    if (u.startsWith('ar://')) {
        return u.replace('ar://', 'https://arweave.net/');
    }

    return u;
}

export const isValidRecipient = (r: string) => {
    if (r === '') {
        return false;
    }

    if (validateStacksAddress(r)) {
        return true;
    }

    try {
        contractPrincipalCVFromString(r);
        return true;
    } catch (e) {

    }

    return false;
}
