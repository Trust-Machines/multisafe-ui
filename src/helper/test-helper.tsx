import {render} from '@testing-library/react';
import {createHistory, createMemorySource, LocationProvider} from '@reach/router';

export const renderWithRouter = (
    ui: any,
    {route = '/', history = createHistory(createMemorySource(route))} = {},
) => {
    return {
        ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
        // adding `history` to the returned utilities to allow us
        // to reference it in our tests (just try to avoid using
        // this to test implementation details).
        history,
    }
}