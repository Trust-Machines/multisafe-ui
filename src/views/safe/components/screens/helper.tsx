import React from 'react';
import Typography from '@mui/material/Typography';
import {Box, useTheme} from '@mui/material';

const ScreenHeader = (props: { title: string, icon: JSX.Element, children: JSX.Element }) => {
    const theme = useTheme();
    const icon = React.cloneElement(props.icon, {
        sx: {
            marginRight: '4px',
            color: theme.palette.primary.main
        }
    })
    return <Box sx={{
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'space-between',
        mb: '20px'
    }}>
        <Typography variant="h6" sx={{display: 'flex', alignItems: 'center',}}>
            {icon} {props.title}
        </Typography>
        {props.children}
    </Box>
}

export default ScreenHeader;