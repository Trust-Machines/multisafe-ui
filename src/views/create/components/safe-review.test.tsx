import SafeReview from './safe-review';
import {renderWithRouter} from '../../../helper/test-helper';

jest.mock('@download/blockies', () => {
    const original = jest.requireActual('@micro-stacks/react');
    return {
        ...original,
        createIcon: () => {
            return {
                toDataURL: () => 'data:image/png;base64,iVBORw0...'
            }
        }
    };
});

test('1 Default render', () => {
    const view = renderWithRouter(<SafeReview name="my-safe-2"
                                        owners={['SP3XD84X3PE79SHJAZCDW1V5E9EA8JSKRBPEKAEK7', 'SP2DXHX9Q844EBT80DYJXFWXJKCJ5FFAX50CQQAWN']}
                                        confirmations={2}
                                        onBack={() => {}}
                                        onNext={() => {}}
                                        onConflict={() => {}}
    />);
    expect(view.container).toMatchSnapshot()
});
