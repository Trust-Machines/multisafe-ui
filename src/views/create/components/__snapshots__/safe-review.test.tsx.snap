// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 Default render 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-156i07"
    >
      <div
        class="review-section MuiBox-root css-1kp2kwa"
      >
        <div
          class="review-section-title"
        >
          Details
        </div>
        <div
          class="detail-section"
        >
          <div
            class="section-title"
          >
            Name of new Safe
          </div>
          <div
            class="section-info"
          >
            my-safe-2
          </div>
        </div>
        <div
          class="detail-section"
        >
          <div
            class="section-title"
          >
            Confirmation threshold
          </div>
          <div
            class="section-info"
          >
            2 out of 2 owners
          </div>
        </div>
      </div>
      <div
        class="review-section MuiBox-root css-1r47yf"
      >
        <div
          class="review-section-title"
        >
          2 Safe owners
        </div>
        <div
          class="MuiBox-root css-1ethbx3"
        >
          <div
            class="icon"
          >
            <img
              alt="SP3XD84X3PE79SHJAZCDW1V5E9EA8JSKRBPEKAEK7"
              src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAoCAYAAACM/rhtAAAABmJLR0QA/wD/AP+gvaeTAAAA5klEQVRYhe2YsQ3CQAxF7xBroAgxAAUMkAGoWIINGCI9U9AyRigYAVHS0oc28UlnWV8n/eK/7iL78vVl2U7y995Pacbx+pkf0zhsUg0bb4nm2/hVNZsACUTJ3XY31QK8GkLxapjeQQlEWdsHtuYut0dTAeNwWpxtTdI7KIEo2c7iKOgs9qB3UAJRij4Y3QdRtA+2hl5geB/0+p5H9D56ByUQxd0HLYfuBb5yuf95NUnvoASi5HP/W/TB53tfTYj2sWi8rXF6ByUQpfgmic5a9P+hdx+9gxKIUsxii9fH0BrTPtgaeoF/SJJL3B6Nz78AAAAASUVORK5CYII="
              style="border-radius: 50%;"
            />
          </div>
          <div
            class="address"
          >
            SP3XD84X3PE79SHJAZCDW1V5E9EA8JSKRBPEKAEK7
          </div>
          <a
            class="link"
            href="https://explorer.stacks.co/address/SP3XD84X3PE79SHJAZCDW1V5E9EA8JSKRBPEKAEK7?chain=mainnet"
            rel="noreferrer"
            target="_blank"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1jm6t9j-MuiSvgIcon-root"
              data-testid="LaunchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
              />
            </svg>
          </a>
        </div>
        <div
          class="MuiBox-root css-1ethbx3"
        >
          <div
            class="icon"
          >
            <img
              alt="SP2DXHX9Q844EBT80DYJXFWXJKCJ5FFAX50CQQAWN"
              src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACgAAAAoCAYAAACM/rhtAAAABmJLR0QA/wD/AP+gvaeTAAAA2klEQVRYhWOUV1T+z4AEznTJMpACTMoe45Wn1DwmknQPABh1IKWA8fVqh//4FFCaxijVP+hDcNSBlAIWQgqM5S+i8Gdm+5BkAXoaS5+6BU0FfvMGfQiOOpBSwIJeTqGnmbMP9dG0kFa3ogNC5o3WxdQGg96BLKS219ABtduD6OoHfQiOOpBSwDjaJ6EQjDqQUkCwT4IOCKU5QoDUNDnoQ3DUgZQCjPYgOiCUZijtFxOSH/QhOOpASsGoAykFow6kFDAGOXxBqYsx+62oAH2shtbqB30IjjqQUgAAhphFHUVHqC4AAAAASUVORK5CYII="
              style="border-radius: 50%;"
            />
          </div>
          <div
            class="address"
          >
            SP2DXHX9Q844EBT80DYJXFWXJKCJ5FFAX50CQQAWN
          </div>
          <a
            class="link"
            href="https://explorer.stacks.co/address/SP2DXHX9Q844EBT80DYJXFWXJKCJ5FFAX50CQQAWN?chain=mainnet"
            rel="noreferrer"
            target="_blank"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1jm6t9j-MuiSvgIcon-root"
              data-testid="LaunchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1bht99n"
    >
      You're about to create a new Safe on 
      <strong>
        Mainnet
      </strong>
       and will have to confirm a transaction with your wallet. The creation cost will be determined by your wallet.
    </div>
    <div
      class="MuiBox-root css-4jc3c7"
    >
      <button
        class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1i3558q-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Back
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Connect Wallet
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
